File: packages/mirascript/src/vm/operations/array-range.ts
import { isFinite } from '../../helpers/utils.js';
import { VM_ARRAY_MAX_LENGTH } from '../../helpers/constants.js';
import type { VmAny } from '../types/index.js';
import { $ToNumber } from './convert.js';

const makeArray = (start: VmAny, end: VmAny, exclusive: boolean): number[] => {
    const s = $ToNumber(start);
    const e = $ToNumber(end);
    if (!isFinite(s) || !isFinite(e) || s > e) {
        return [];
    }
    const n = exclusive ? Math.ceil(e - s) : Math.floor(e - s + 1);
    if (n > VM_ARRAY_MAX_LENGTH) {
        throw new RangeError(`Array length exceeds maximum limit of ${VM_ARRAY_MAX_LENGTH}`);
    }
    const arr = [];
    for (let i = 0; i < n; i++) {
        arr[i] = s + i;
    }
    return arr;
};
/** 构造范围数组 */
export const $ArrayRange = (start: VmAny, end: VmAny): number[] => {
    return makeArray(start, end, false);
};
/** 构造范围数组（不包含结束值） */
export const $ArrayRangeExclusive = (start: VmAny, end: VmAny): number[] => {
    return makeArray(start, end, true);
};

File: crates/python/mirascript/_vm/operations/array_range.py
from __future__ import annotations
import math

from ..._helpers.constants import VM_ARRAY_MAX_LENGTH
from ..types import VmAny
from .convert import ToNumber


def _make_array(start: VmAny, end: VmAny, exclusive: bool) -> list[float]:
    s = ToNumber(start)
    e = ToNumber(end)
    if not math.isfinite(s) or not math.isfinite(e) or s > e:
        return []
    n = math.ceil(e - s) if exclusive else math.floor((e - s) + 1.0)
    if n > VM_ARRAY_MAX_LENGTH:
        raise RuntimeError(
            f"Array length exceeds maximum limit of {VM_ARRAY_MAX_LENGTH}"
        )
    arr: list[float] = [0.0] * n
    for i in range(n):
        arr[i] = s + i
    return arr


def ArrayRange(start: VmAny, end: VmAny) -> list[float]:
    return _make_array(start, end, exclusive=False)


def ArrayRangeExclusive(start: VmAny, end: VmAny) -> list[float]:
    return _make_array(start, end, exclusive=True)
